python/pyarrow/_s3fs.pyx

```diff
@@ -91,85 +97,40 @@ cdef class S3Options:
             )
         else:
-            self.options = CS3Options.Defaults()
+            options = CS3Options.Defaults()
 
         if region is not None:
-            self.region = region
+            options.region = tobytes(region)
         if scheme is not None:
-            self.scheme = scheme
+            options.scheme = tobytes(scheme)
         if endpoint_override is not None:
-            self.endpoint_override = endpoint_override
+            options.endpoint_override = tobytes(endpoint_override)
         if background_writes is not None:
-            self.background_writes = background_writes
-
-    cdef inline CS3Options unwrap(self) nogil:
-        return self.options
-
-    @property
-    def region(self):
-        """AWS region to connect to."""
-        return frombytes(self.options.region)
-
-    @region.setter
-    def region(self, value):
-        self.options.region = tobytes(value)
-
-    @property
-    def scheme(self):
-        """S3 connection transport scheme."""
-        return frombytes(self.options.scheme)
-
-    @scheme.setter
-    def scheme(self, value):
-        self.options.scheme = tobytes(value)
-
-    @property
-    def endpoint_override(self):
-        """Override region with a connect string such as localhost:9000"""
-        return frombytes(self.options.endpoint_override)
-
-    @endpoint_override.setter
-    def endpoint_override(self, value):
-        self.options.endpoint_override = tobytes(value)
-
-    @property
-    def background_writes(self):
-        """OutputStream writes will be issued in the background"""
-        return self.options.background_writes
-
-    @background_writes.setter
-    def background_writes(self, bint value):
-        self.options.background_writes = value
+            options.background_writes = background_writes
 
-
-cdef class S3FileSystem(FileSystem):
-    """S3-backed FileSystem implementation
-
-    Note: S3 buckets are special and the operations available on them may be
-    limited or more expensive than desired.
-
-    Parameters
-    ----------
-    options: S3Options, default None
-        Options for connecting to S3. If None is passed then attempts to
-        initialize the connection from AWS environment variables.
-    """
-
-    cdef:
-        CS3FileSystem* s3fs
-
-    def __init__(self, S3Options options=None):
-        cdef shared_ptr[CS3FileSystem] wrapped
-        options = options or S3Options()
         with nogil:
-            wrapped = GetResultValue(CS3FileSystem.Make(options.unwrap()))
+            wrapped = GetResultValue(CS3FileSystem.Make(options))
+
         self.init(<shared_ptr[CFileSystem]> wrapped)
 
     cdef init(self, const shared_ptr[CFileSystem]& wrapped):
         FileSystem.init(self, wrapped)
         self.s3fs = <CS3FileSystem*> wrapped.get()
+
+    def __reduce__(self):
+        cdef CS3Options opts = self.s3fs.options()
+        return (
+            S3FileSystem, (
+                None,  # frombytes(opts.access_key),
```

Comment:
@pitrou I'm not sure how to retrieve the credentials. Pickling the credentials could also cause security issues, would be an explicit notice enough?